### DISCUSSION TITLE
Custom MQTT transport layer

### DISCUSSION THREAD

**Discussion by edorgeville** [trpc/trpc#3639] [Q&A] [answered]
Hi,
We are looking at tRPC to use as a service-to-service communication layer. It seems like a very good candidate that would greatly simplify our development process as well as testing.  
One thing that I would like to achieve is to use MQTT as a transport layer instead of HTTP / Websocket.
For the client-side, I did have look at the ["Creating a custom link"](https://trpc.io/docs/links#creating-a-custom-link) section of the docs. I've been trying to wrap my head around this section, without too much results yet, but I feel that with enough time I'll get there. But what about server-side? How should I approach this?
If I end up making both server-side & client-side implementations, is there a suggested way to package such a "transport plugin" in order to contribute here? Is it a feature that the tRPC team would be interested in merging or should I maintain a seperate npm package?
Thanks in advance for answers and pointers :^]
Erwan

**Comment by edorgeville** [comment, marked as answer]
I started work on an [MQTT adapter for tRPC](https://github.com/edorgeville/trpc-mqtt), based off of @imxeno's [RabbitMQ adapter](https://github.com/imxeno/trpc-rabbitmq). I got a proof-of-concept working today, hopefully I can finish, clean up & publish this soon™!  
EDIT: module is published and functional :)

**Reply by KATT** [reply]
Ohhh cool 🍿

**Reply by ghost** [reply]
can you publish your PoC?

**Reply by edorgeville** [reply]
@sargtier1-rs the npm package is [published here](https://www.npmjs.com/package/trpc-mqtt), and the source [here](https://github.com/edorgeville/trpc-mqtt) :)

**Reply by ghost** [reply]
merci!
